test/src/us/kbase/kbasefeaturevalues/test/KBaseFeatureValuesServerTest.java

```diff
@@ -195,38 +197,46 @@ public void testMainPipeline() throws Exception {
         String clustObj3Name = "clusters3";
         ExpressionMatrix data = new ExpressionMatrix().withType("log-ratio").withScale("1.0")
                 .withData(getSampleMatrix());
-        wscl.saveObjects(new SaveObjectsParams().withWorkspace(testWsName).withObjects(Arrays.asList(
+        getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName).withObjects(Arrays.asList(
                 new ObjectSaveData().withName(exprObjName).withType("KBaseFeatureValues.ExpressionMatrix")
                 .withData(new UObject(data)))));
-        //Assert.assertNotNull("exprObjName ",testWsName + "/" + exprObjName);
-        //Assert.assertEquals("exprObjName ",testWsName + "/" + exprObjName);
-        //Assert.assertNotNull("estimObjName ", estimObjName);
-        //
         /////////////// estimate K /////////////////
         EstimateKParams ekp = new EstimateKParams().withInputMatrix(testWsName + "/" +
                 exprObjName).withOutWorkspace(testWsName).withOutEstimateResult(estimObjName);
-        impl.estimateK(ekp, token, getContext());
-        ObjectData res1 = wscl.getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
-                .withName(estimObjName))).get(0);
-        EstimateKResult estKRes = res1.getData().asClassInstance(EstimateKResult.class);
+        EstimateKResult estKRes = impl.estimateK(ekp, token, getContext());
+
+        //ObjectData res1 = getWsClient().getObjects(Arrays.asList(new ObjectIdentity().withWorkspace(testWsName)
+        //        .withName(estimObjName))).get(0);
+        //EstimateKResult estKRes = res1.getData().asClassInstance(EstimateKResult.class);
         long k = estKRes.getBestK();
         Assert.assertNotNull("k exists ", k);
         Assert.assertNotNull("c size exists ", estKRes.getEstimateClusterSizes().size());
-	//System.err.println("k "+k);
-        //System.err.println("cluster size "+estKRes.getEstimateClusterSizes().size());
         Assert.assertEquals(3, k);
         Assert.assertEquals(5, estKRes.getEstimateClusterSizes().size());
         for (int i = 0; i < estKRes.getEstimateClusterSizes().size(); i++) {
             Tuple2 <Long, Double> item = estKRes.getEstimateClusterSizes().get(i);
             Assert.assertEquals(2L + i, (long)item.getE1());
             Assert.assertTrue((double)item.getE2() > 0);
         }
-        impl.estimateKNew(new EstimateKParamsNew().withInputMatrix(testWsName + "/" +
+
+	List<Tuple11<Long,String,String,String,Long,String,Long,String,String,Long,Map<String,String>>> info2 = getWsClient().saveObjects(new SaveObjectsParams().withWorkspace(testWsName)
```

Comment:
I don't understand this. You run the method and get the return result. You then save that result to the workspace, pull it back, and then compare that object to the object returned by the impl result. I'm not sure what this tests.

I would expect that you manually construct an expected result for `estKRes` and then compare `extKRes` to that expected result.

Reply:
Also you're only checking a portion of the returned object instead of the entire object, it seems.

Reply:
If the method also saves the result to the workspace as well as returning it, then the test should pull that result, not save a new object.

Reply:
indeed, there were a mix of cases for the returns -- this test only applied to returns which were a string ref. removed now.